File: 1008-binary-tree-cameras/binary-tree-cameras.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minCameraCover(self, root: Optional[TreeNode]) -> int:
        self.count=0
        def traversal(root):
            if root==None:
                return
            if root.left==None and root.right==None:
                return 'leaf'
            left=traversal(root.left)
            right=traversal(root.right)

            if left=='leaf' or right=='leaf':
                self.count+=1
                return 'camera'
            if left=='camera' or right=='camera':
                return 'gap'
            if left=='gap' or right=='gap':
                return 'leaf'
        #---------------------------------
        if traversal(root)=='leaf':
            self.count+=1
        return self.count

File: 1008-binary-tree-cameras/binary-tree-cameras.java

class Solution {
    int count=0;
    public String dfs(TreeNode root){
        if(root==null){
            return "";
        }
        if(root.left==null && root.right==null){
            return "leaf";
        }
        String left=dfs(root.left);
        String right=dfs(root.right);

        if(left.equals("leaf") || right.equals("leaf")){
            count++;
            return "camera";
        }
        if(left.equals("camera") || right.equals("camera")){
            return "gap";
        }
        return "leaf";

    }
    public int minCameraCover(TreeNode root) {
        if(dfs(root).equals("leaf")){
            count++;
        }
        return count;
    }
}